File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    vector<vector<int>> allpath_;
    vector<int> path_;
public:
    void traverse(TreeNode* node, int rest){
        if (!node->left && !node->right && rest == node->val){
            allpath_.push_back(path_);
            return;
        }
        if (!node->left && !node->right) return;
       
        if (node->left) {
            path_.push_back(node->left->val);
            traverse(node->left, rest-node->val);
            path_.pop_back();
        }
        if (node->right) {
            path_.push_back(node->right->val);
            traverse(node->right, rest-node->val);
            path_.pop_back();
        }
        return;
    }
    
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        allpath_ = vector<vector<int>>();
        if (root == nullptr)
            return allpath_;
        path_ = {root->val};
        traverse(root, targetSum);
        return allpath_;
    }
};

File: 0113-path-sum-ii/0113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.allpath_ = []
        self.path_ = []
    def findPath(self, node, rest):
        if node is None:
            return
        self.path_.append(node.val)
        if node.left is None and node.right is None and rest == node.val:
            self.allpath_.append(list(self.path_))
        else:
            self.findPath(node.left, rest - node.val)
            self.findPath(node.right, rest - node.val)
        self.path_.pop()
        return
    
    def pathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        self.findPath(root, targetSum)
        return self.allpath_
        